File: backend/app/domain/services/configuration.py
"""
Configuration Service.

Manages application configuration settings.
"""

from sqlalchemy.orm import Session
from sqlalchemy import text
from app.domain.repositories.configuration_repo import ConfigurationRepository
from app.domain.schemas.configuration import WALThresholds, BatchConfiguration


class ConfigurationService:
    """Service for managing configuration settings."""
    
    def __init__(self, db: Session):
        """Initialize service."""
        self.repo = ConfigurationRepository(db)
    
    def get_wal_thresholds(self) -> WALThresholds:
        """
        Get WAL monitoring thresholds from configuration.
        
        Note: Thresholds in database are stored in MB and converted to bytes here.
        
        Returns:
            WAL thresholds configuration with values in bytes
        """
        # Get values in MB from database
        warning_mb = int(self.repo.get_value('WAL_MONITORING_THRESHOLD_WARNING', '3000'))
        error_mb = int(self.repo.get_value('WAL_MONITORING_THRESHOLD_ERROR', '6000'))
        webhook_url = self.repo.get_value('ALERT_NOTIFICATION_WEBHOOK_URL', '')
        notification_iteration = int(self.repo.get_value('NOTIFICATION_ITERATION_DEFAULT', '3'))
        
        # Convert MB to bytes (1 MB = 1024 * 1024 bytes)
        warning_bytes = warning_mb * 1024 * 1024
        error_bytes = error_mb * 1024 * 1024
        
        return WALThresholds(
            warning=warning_bytes,
            error=error_bytes,
            webhook_url=webhook_url,
            notification_iteration=notification_iteration
        )
    
    def get_value(self, key: str, default: str = "") -> str:
        """
        Get configuration value by key.
        
        Args:
            key: Configuration key
            default: Default value if not found
            
        Returns:
            Configuration value
        """
        return self.repo.get_value(key, default)
    
    def set_value(self, key: str, value: str):
        """
        Set configuration value.
        
        Args:
            key: Configuration key
            value: Configuration value
        """
        return self.repo.set_value(key, value)
    
    def get_batch_configuration(self) -> BatchConfiguration:
        """
        Get batch configuration settings.
        
        Returns:
            Batch configuration with max_batch_size and max_queue_size
        """
        max_batch_size = int(self.repo.get_value('PIPELINE_MAX_BATCH_SIZE', '4096'))
        max_queue_size = int(self.repo.get_value('PIPELINE_MAX_QUEUE_SIZE', '16384'))
        
        return BatchConfiguration(
            max_batch_size=max_batch_size,
            max_queue_size=max_queue_size
        )
    
    def update_batch_configuration(self, config: BatchConfiguration) -> BatchConfiguration:
        """
        Update batch configuration settings and mark all pipelines for refresh.
        
        Args:
            config: New batch configuration values
            
        Returns:
            Updated batch configuration
        """
        self.repo.set_value('PIPELINE_MAX_BATCH_SIZE', str(config.max_batch_size))
        self.repo.set_value('PIPELINE_MAX_QUEUE_SIZE', str(config.max_queue_size))
        
        # Mark all pipelines to be restarted with new configuration
        self.repo.db.execute(
            text("UPDATE pipelines SET ready_refresh = TRUE WHERE status = 'START'")
        )
        self.repo.db.commit()
        
        return config

File: web/src/repo/configuration.ts
import { api } from './client'

export interface WALThresholdsConfig {
    warning: number  // in MB
    error: number    // in MB
    enable_webhook: boolean
    webhook_url: string
    notification_iteration: number
    enable_telegram: boolean
    telegram_bot_token: string
    telegram_chat_id: string
}

export interface BatchConfiguration {
    max_batch_size: number
    max_queue_size: number
}

export const configurationRepo = {
    getWALThresholds: async () => {
        const { data } = await api.get<WALThresholdsConfig>('/configuration/wal-thresholds')
        return data
    },

    updateWALThresholds: async (config: WALThresholdsConfig) => {
        const { data } = await api.put<WALThresholdsConfig>('/configuration/wal-thresholds', config)
        return data
    },

    testNotification: async (webhook_url?: string, telegram_bot_token?: string, telegram_chat_id?: string) => {
        const { data } = await api.post('/configuration/wal-thresholds/test', { 
            webhook_url,
            telegram_bot_token,
            telegram_chat_id
        })
        return data
    },

    getBatchConfiguration: async () => {
        const { data } = await api.get<BatchConfiguration>('/configuration/batch')
        return data
    },

    updateBatchConfiguration: async (config: BatchConfiguration) => {
        const { data } = await api.put<BatchConfiguration>('/configuration/batch', config)
        return data
    },
}
